javascript/documentation/library-customization.rst

```diff
@@ -0,0 +1,249 @@
+``ASTNode.getChild`` to change the way the AST structure is represented, but this should normally be
+avoided in the interest of keeping a close correspondence between AST and concrete syntax.
+
+CFG
+~~~
+
+You can override ``ControlFlowNode.getASuccessor`` to customize the control-flow graph. Note that overriding ``ControlFlowNode.getAPredecessor`` is not normally useful, since it is rarely used in higher-level libraries.
+
+SSA
+~~~
+
+It is not normally necessary to customize this layer.
+
+Local data flow
+~~~~~~~~~~~~~~~
+
+The ``DataFlow::SourceNode`` class uses the range pattern, so new kinds of source nodes can be
+added by extending ``Dataflow::SourceNode::Range``. Some of its subclasses can similarly be
+extended: ``DataFlow::ModuleImportNode`` models module imports, and ``DataFlow::ClassNode`` models
+class definitions. The former provides default implementations covering CommonJS, AMD and ECMAScript
+2015 modules, while the latter handles ECMAScript 2015 classes as well as traditional function-based
+classes. You can extend their corresponding ``::Range`` classes to add support for other module or
+class systems.
+
+Type inference
+~~~~~~~~~~~~~~
+
+You can override ``AnalyzedNode::getAValue`` to customize the type inference. Note that the type inference is expected to be sound, that is (as far as practical) the abstract values inferred for a data-flow nodes should cover all possible concrete values this node may take on at runtime.
+
+You can also extend the set of abstract values in one of two ways:
+
+1. To add individual abstract values that are independent of the program being analyzed, define a
+   subclass of ``CustomAbstractValueTag`` describing the new abstract value. There will then be a
+   corresponding value of class ``CustomAbstractValue`` that you can use in overriding
+   definitions of the ``getAValue`` predicate.
+2. To add abstract values that are induced by a program element, define a subclass of
+   ``CustomAbstractValueDefinition``, and use its corresponding
+   ``CustomAbstractValueFromDefinition``.
+
+Call graph
+~~~~~~~~~~
+
+You can override ``DataFlow::InvokeNode::getACallee(int)`` to customize the call graph. Note that overriding the zero-argument version ``getACallee()`` is not enough since higher layers use the one-argument version. 
+
+Type tracking
+~~~~~~~~~~~~~
+
+It is not normally necessary to customize this layer.
+
+Framework models
+~~~~~~~~~~~~~~~~
+
+The ``semmle.javascript.frameworks.HTTP`` module defines many abstract classes that can be extended
+to implement support for new web frameworks. These classes, in turn, are used by some of the
+security queries (such as the cross-site scripting queries) to define sources and sinks, so these
+queries will automatically benefit from the additional modeling.
+
+Similarly, the ``semmle.javascript.frameworks.SQL`` module defines abstract classes for modeling SQL
```

Comment:
Can we add `semmle.javascript.frameworks.ClientRequests` to this list, and perhaps explain the difference between that and the server request handlers of `HTTP.qll`?